### ISSUE TITLE
Associated type constraints on super traits allowing for unsound transmutation to trait objects  

### ISSUE THREAD

**Issue by Kyykek** [rust-lang/rust#114389]
This code allows for unsound transmutation to trait objects of trait with impossible constraints.

```rust
trait TypeEq {
    type This: ?Sized;
}

impl<T: ?Sized> TypeEq for T {
    type This = T;
}

fn identity<T: ?Sized>(t: &<T as TypeEq>::This) -> &T {
    t  // T::This is deduced to be T
}

trait Cat {
    fn meow(&self) {
        println!("meow");
    }
}

struct SomeType;
impl Cat for SomeType {}

// `dyn Dog` isn't a valid type, but we are still allowed to create a trait object of this
trait Dog: TypeEq<This = dyn Cat> {
    fn bark(&self) {
        println!("bark");
    }
}

pub fn main() {
    let normal_coercion: &dyn Cat = &SomeType;

    // `T` is inferred to be `dyn Dog`, `T::This` is inferred as `dyn Cat`, which is different from how the compiler
	// inferred these types in the function definition.
    let oh_no_my_vtable: &dyn Dog = identity(normal_coercion);

    oh_no_my_vtable.bark(); // meow :3
}
```
I'm not sure if this is a special case of one of the numerous similar known bugs, so I thought I should share it just in case.  Simpler cases where the types' memory layouts do not match cause the compiler to panic or llvm to segfault during optimization.

**Comment by asquared31415** [comment]
A trivial change allows a segfault in safe code:
```rust
trait Dog: TypeEq<This = dyn Cat> {
    fn bark(&self) {
        println!("bark");
    }

    fn extra(&self) {
        panic!("extra");
    }
}

// ... 

oh_no_my_vtable.extra(); // segfault
```

**Comment by Kyykek** [comment]
Note that that code does not segfault with rustc 1.71; it simply does nothing. 1.70 compiles it to a segfault.
Both 1.71 and 1.70 compile to a SIGILL when turning on optimizations.

[https://godbolt.org/z/rb9be9vTn](Compiler Explorer)

**Comment by Kyykek** [comment]
Here is the compiler panic example (presumably caused by the transmutation of a thin pointer to a fat pointer)
```rust
trait TypeEq {
    type This: ?Sized;
}

impl<T: ?Sized> TypeEq for T {
    type This = Self;
}

fn identity<T: ?Sized>(t: &<T as TypeEq>::This) -> &T {
    t
}

struct SomeType;

// compiler crashes if called
fn unsound(x: &SomeType) -> &dyn TypeEq<This = SomeType> {
    identity(x)
}

fn main() {
    let thing = SomeType;
    unsound(&thing);  // crash!
}
```
[Compiler Explorer](https://godbolt.org/z/WMz13T5v1)


Here is the example that causes a libLLVM segfault when using -O
```rust
trait TypeEq {
    type This: ?Sized;
}

impl<T: ?Sized> TypeEq for T {
    type This = Self;
}

fn identity<T: ?Sized>(t: &<T as TypeEq>::This) -> &T {
    t
}

trait Dog: TypeEq<This = [u8]> {
    fn woof(&self) {
        println!("woof");
    }
}

fn main() {
    let arr = vec![1; 0];
    let arr: &[u8] = &arr; // fat pointer (ptr, len) with len being 0
    let dog: &dyn Dog = identity(arr);  // (data_ptr, vtable_ptr) with vtable_ptr having address 0
    dog.woof(); // llvm does not like this.
}
```
[Compiler Explorer](https://godbolt.org/z/j58rPqGPh)

Replacing `vec![1; 0]` with `Vec::new()` or just `[1; 0]` instead causes the program to segfault; there's some weird stuff going on with when combined with `std::vec::from_elem()` (the `doc(hidden)` function called by vec! in this case).
The segfault also happened when i first came across this while writing type level magic. The compiler just segfaulting on me without any sort of error message initially made me think that I had screwed up something with my environment.

**Comment by Kyykek** [comment]
This is very similar to #57893, in that an incorrect associated type bound is assumed correct.

**Comment by wwylele** [comment]
This can be exploited without involving super traits. The trick is to construct the impossible type `dyn TypeEq<This = dyn Cat>`

```rust
trait TypeEq {
    type This: ?Sized;
    fn bark(&self) {
        println!("bark");
    }
}

impl<T: ?Sized> TypeEq for T {
    type This = T;
}

fn identity<T: ?Sized>(t: &<T as TypeEq>::This) -> &T {
    t  // T::This is deduced to be T
}

trait Cat {
    fn meow(&self) {
        println!("meow");
    }
}

struct SomeType;
impl Cat for SomeType {}

pub fn main() {
    let normal_coercion: &dyn Cat = &SomeType;

    // `T` is inferred to be `dyn Dog`, `T::This` is inferred as `dyn Cat`, which is different from how the compiler
	// inferred these types in the function definition.
    let oh_no_my_vtable: &dyn TypeEq<This = dyn Cat> = identity(normal_coercion);

    oh_no_my_vtable.bark(); // meow :3
}
```

**Comment by Kyykek** [comment]
Well, the fundamental issue at hand seems to be that if we say that `dyn Tr` automatically implements `Tr`, then inferring the type `This` from the `impl` body is simply unsound because it doesnt take into account all possible types that `This` can take on.

`impl<T: ?Sized> Tr for T` does not implement `Tr` for `dyn Tr` because that would be a conflicting implementation, as can be seen when trying to `impl Tr for dyn Tr`. As such, `dyn Tr<This=Impossible>` does not contradict the implementation, because it doesn't apply to it. 

Coercing `SomeType` into `dyn TypeEq<This=SomeType>` is unsound because the only reason that `SomeType` implements 
`TypeEq<This=SomeType>` is that it fulfills `T == SomeType` for its implementation which is not the case for `dyn TypeEq<This=SomeType>`.

The underlying issue is that by inferring the associated type from the blanket implementation we assume that it applies to all types, when it simply does not because `dyn Tr` works differently. 

If we want to infer types from `impl` blocks because "the implementation is always the same" it's not enough to just look at the impl.

```rust
fn identity<T: ?Sized>(t: &<T as TypeEq>::This) -> &T {
    t
}
```
I don't think this function should compile, since there are types that satisfy the trait bounds of `T` for which inferring that `This == T` is not correct.

In my eyes, #57893 and this have the same root cause, but I do not think that normalization as mentioned over there is enough to fully fix this issue.

**Comment by Kyykek** [comment]
Alternatively, one could require something like that to be object safe (for a given equality bound), a trait object type like `dyn TypeEq<This=T>` needs to fulfill any implementation (i.e. any blanket implementation) whose conditions apply to the trait object type.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23114389.20Associated.20type.20constraints.20on.20super.20traits.20allowin.E2.80.A6)).

I'm going to nominate this issue for the next T-compiler meeting to get more eyeballs on this (and the related old issue mentioned). I'd like to understand better what's the current status of this unsoundness, how we can proceed and in which timeframe.

@rustbot label +I-compiler-nominated

**Comment by Kyykek** [comment]
I would also like to remind that more complex bounds might make restricting object safety without breaking sound code complicated.
```rust
trait Tr {
   type T: ?Sized;
}

trait ExtraBound {}

impl<T: ?Sized+ExtraBound> Tr for T {
   type T=T;
}

trait TrExtra: Tr+ExtraBound {}
impl<T: Tr+ExtraBound+?Sized> TrExtra for T {}
// this is sound since the blanket impl required ExtraBound.
type Sound = dyn Tr<T=i32>;
type Unsound = dyn TrExtra<T=i32>;
```
Though you could argue that the blanket impl for `TrExtra` causes it to not be object safe, while `Tr` is.

This gets more interesting with `dyn Tr+SomeAutoTrait`
```rust
// second example with more complex trait obj 
trait Tr {
   type T: ?Sized;
}
impl<T: ?Sized+Send> Tr for T {
   type T=T;
}
// no contradiction with blanket because no Send
type Sound = dyn Tr<T=i32>;
// but this is not sound even though both its components are
type Unsound = dyn Tr<T=i32>+Send;
```

Also `dyn Tr+'lifetime`?
```rust
trait Tr {
   type T: ?Sized;
}
impl<T: ?Sized+'static> Tr for T {
   type T=T;
}

type MightBeSound = dyn Tr<T=i32>;
type Unsound = dyn Tr<T=i32>+'static;
```

I do not even want to think about how something like
```rust
trait Tr<'a>: 'a {}
```
would play out here.

**Comment by Kyykek** [comment]
I saw some of the discussion on zulip, #57893 is still completely broken. I have tried most of the examples there and all I've tried still worked.
Also I can confirm that I have actually run into this on accident.
Furthermore, I'm 100% sure that this is just a special, more practical case of #57893. The core issue is the lack of checking of blanket implementations that could but do not apply to the `dyn Tr` type when inferring associated types.
